File: BackEnd/app/models/kiosk.py
import uuid
from datetime import datetime
from sqlalchemy import String, ForeignKey, TIMESTAMP
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID, JSONB

from app.db.base import Base


class KioskDevice(Base):
    __tablename__ = "kiosk_devices"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    tenant_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False
    )
    device_code: Mapped[str] = mapped_column(String(50), nullable=False)
    name: Mapped[str | None] = mapped_column(String(255))
    location: Mapped[str | None] = mapped_column(String(255))
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="online")
    last_heartbeat: Mapped[datetime | None] = mapped_column(TIMESTAMP(timezone=True))
    config: Mapped[dict] = mapped_column(
        JSONB, nullable=False, default=dict, server_default="{}"
    )
    created_at: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True), default=datetime.utcnow
    )
    updated_at: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True), default=datetime.utcnow
    )

    tenant = relationship("Tenant", back_populates="kiosk_devices")
    sessions = relationship("KioskSession", back_populates="device")
    bookings = relationship("Booking", back_populates="device")


class KioskSession(Base):
    __tablename__ = "kiosk_sessions"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    tenant_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False
    )
    device_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("kiosk_devices.id"), nullable=True
    )
    session_token: Mapped[str] = mapped_column(String(255), unique=True, nullable=False)
    language: Mapped[str] = mapped_column(String(10), nullable=False, default="en")
    started_at: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True), default=datetime.utcnow
    )
    ended_at: Mapped[datetime | None] = mapped_column(TIMESTAMP(timezone=True))
    final_state: Mapped[str | None] = mapped_column(String(50))
    created_at: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True), default=datetime.utcnow
    )

    device = relationship("KioskDevice", back_populates="sessions")
    bookings = relationship("Booking", back_populates="session")

File: FrontEnd/domain/entities/Kiosk.ts
// Kiosk domain entities — matches kiosk_devices and kiosk_sessions in code.sql

export type KioskStatus = 'online' | 'offline' | 'maintenance';
export type KioskFinalState = 'COMPLETE' | 'IDLE' | 'ERROR';

export interface KioskDevice {
    id: string;
    tenantId: string;
    deviceCode: string;            // mapped from device_code (unique per tenant)
    name?: string;
    location?: string;
    status: KioskStatus | string;  // default: 'online'
    lastHeartbeat?: string;        // mapped from last_heartbeat
    config: Record<string, unknown>; // JSONB
    createdAt?: string;
    updatedAt?: string;
}

export interface KioskSession {
    id: string;
    tenantId: string;
    deviceId?: string;             // mapped from device_id
    sessionToken: string;          // mapped from session_token (unique)
    language: string;              // default: 'en'
    startedAt: string;             // mapped from started_at
    endedAt?: string;
    finalState?: KioskFinalState | string;
    createdAt?: string;
}
